### PR TITLE
Add --wait flag to svcat commands

### PR DESCRIPTION
**Pull request by carolynvs**
Waits for the operation to complete (or fail). Use in conjunction with the `--timeout 30s` flag to limit how long svcat will wait before giving up. Timeout is a human readable golang duration, like 1us, 20ms, 50ns, 100s, 2m, 1h.

```
$ svcat provision --wait

$ svcat bind --wait

$ svcat deprovision --wait

#  waits for all bindings associated with the instance to be deleted.
$ svcat unbind INSTANCE

$ svcat unbind --name BINDING --wait
```
I went back and forth on if wait should be the default or not. Ultimately I decided that it shouldn't because that's inconsistent with kubectl, and for the most part other clis as well don't wait by default. Also it seems more clear when you say `--wait --timeout 5m` instead of just a timeout by itself which may make you think that it applies to the operation and not the wait.

Part 1 of #1736 that covers the following svcat commands: provision, deprovision, bind and unbind.

### CODE REVIEW

**Comment by carolynvs** [comment]
I want to tackle next moving the svcat tests to our e2e tests so that they hit a real cluster instead of the test http server. I think it will be easier to test stuff like this (and the other mutating verbs) once we are moved over.

**Comment by carolynvs** [comment]
Right now this is how the svcat tests work:

* I recorded the json body returned from the api server for requests like "get serviceclasses" and saved them in cmd/svcat/testdata/responses
* The svcat tests have a test httpserver than converts the requested url into a path to the recorded response in that directory and returns the file contents.
* The tests then compare the output from the cli to the golden file output in cmd/svcat/testdata/output.

I've gotten a bit of flack over how this all works (both when I first donated the code and later). People have said that they would prefer that the tests either use the fake clients (I didn't know that existed at the time), or were changed to e2e tests that actually hit the binary from the shell. A big part of that is that setting up new testdata or updating it is not straightforward.

Since the tests all work against a live ups-broker (that's the testdata that I recorded), I figured it would be easy to change the tests to run as e2e tests and also add tests for verbs that mutate state (which we don't test right now at all).

That was a long explanation, maybe the discussion should be moved into the existing issue https://github.com/kubernetes-incubator/service-catalog/issues/1810? 😀

**Comment by carolynvs** [comment]
As a side note, I'm making a kubectl plugin that will help you do the same thing but for pods and deployments. ✨

File: cmd/svcat/binding/bind_cmd.go [hunk 1]
```diff
@@ -126,6 +143,12 @@ func (c *bindCmd) bind() error {
 		return err
 	}
 
+	if c.wait {
+		glog.V(2).Info("Waiting for instance to be bound...")
```
Comment (jboyd01) [inline cmd/svcat/binding/bind_cmd.go]: nit, do with this as you wish:  I dislike the idea of an instance being bound.  It sounds singular to me, I prefer the notion of creating a new binding.  If you don't agree, ignore my comment :)

File: pkg/svcat/service-catalog/secret.go [hunk 2]
```diff
@@ -29,14 +29,10 @@ import (
...
 func (sdk *SDK) RetrieveSecretByBinding(binding *v1beta1.ServiceBinding) (*corev1.Secret, error) {
-	cond := GetBindingStatusCondition(binding.Status)
-	isReady := cond.Type == v1beta1.ServiceBindingConditionReady &&
-		cond.Status == v1beta1.ConditionTrue
-
 	secret, err := sdk.Core().Secrets(binding.Namespace).Get(binding.Spec.SecretName, metav1.GetOptions{})
 	if err != nil {
 		// It's expected to not have the secret until the binding is ready
-		if !isReady && errors.IsNotFound(err) {
+		if !sdk.IsBindingReady(binding) && errors.IsNotFound(err) {
```
Comment (jboyd01) [inline pkg/svcat/service-catalog/secret.go]: These changes to RetrieveSecretByBinding appear to be  unrelated to --wait, but they look intention and must be fixing issues?

Comment (carolynvs) [reply pkg/svcat/service-catalog/secret.go]: I realized that I needed the exact logic from this function in 3 places (is the binding failed or ready?), so I made a function and used it from this spot, in addition to in the wait function and the cli.

**Comment by carolynvs** [comment]
This is what it looks like when the timeout is hit before the operation is complete

```console
$ svcat provision iloveponies --class mysql --plan 5-7-14 --wait --timeout 1ns
  Name:        iloveponies
  Namespace:   default
  Status:
  Class:       mysql
  Plan:        5-7-14

Parameters:
  {}
Error: timed out waiting for the condition
```

File: cmd/svcat/binding/bind_cmd.go (hunk 1 above)
Comment (carolynvs) [reply cmd/svcat/binding/bind_cmd.go]: I changed it to `Waiting for binding to be injected...`. I can't say "binding to be created" because it exists immediately, and we are really just waiting for it be usable.

**Comment by jboyd01** [comment]
LGTM, svcat continues to rock!

File: cmd/svcat/binding/bind_cmd.go [hunk 3]
```diff
@@ -126,6 +143,12 @@ func (c *bindCmd) bind() error {
...
 
+	if c.wait {
+		glog.V(2).Info("Waiting for binding to be injected...")
+		pollInterval := 1 * time.Second
+		binding, err = c.App.WaitForBinding(binding.Namespace, binding.Name, pollInterval, c.timeout)
+	}
+
 	output.WriteBindingDetails(c.Output, binding)
-	return nil
+	return err
```
Comment (n3wscott) [inline cmd/svcat/binding/bind_cmd.go]: This deserves a comment that it is assumed err will be either nil from c.App.Bind or [nil or err] from c.App.WaitForBinding.

I would vote the more verbose way of using the pattern of if err ! nil, return err inside of c.wait and revert this change and make it clear that this return means bind() executed clean.

File: cmd/svcat/binding/unbind_cmd.go [hunk 4]
```diff
@@ -82,6 +118,33 @@ func (c *unbindCmd) deleteBinding() error {
...
 	bindings, err := c.App.Unbind(c.Namespace, c.instanceName)
-	output.WriteDeletedBindingNames(c.Output, bindings)
-	return err
+	if err != nil {
+		return err
+	}
+
+	if c.wait {
+		glog.V(2).Infof("Waiting for the bindings to be deleted...")
+		pollInterval := 1 * time.Second
```
Comment (n3wscott) [inline cmd/svcat/binding/unbind_cmd.go]: I would recommend making a flag to be able to set the poll interval and the default can still be 1 second. A 1 second poll delay in a large script could mean it runs significantly slower if the commands finish in 1001 milliseconds

File: cmd/svcat/instance/deprovision_cmd.go [hunk 5]
```diff
@@ -53,6 +63,14 @@ func (c *deprovisonCmd) Validate(args []string) error {
 	}
 	c.instanceName = args[0]
 
+	if c.wait && c.rawTimeout != "-1" {
```
Comment (n3wscott) [inline cmd/svcat/instance/deprovision_cmd.go]: This type of logic has been repeated a few times, I am wondering if you could make a helper function and pass in an anon function to do the custom work and delegate the time logic to a common place.

File: cmd/svcat/instance/deprovision_cmd.go [hunk 6]
```diff
@@ -62,6 +80,23 @@ func (c *deprovisonCmd) Run() error {
 
 func (c *deprovisonCmd) deprovision() error {
 	err := c.App.Deprovision(c.Namespace, c.instanceName)
+	if err != nil {
+		return err
+	}
+
+	if c.wait {
+		glog.V(2).Infof("Waiting for the instance to be deleted...")
```
Comment (n3wscott) [inline cmd/svcat/instance/deprovision_cmd.go]: Same helper function idea here.

File: cmd/svcat/instance/provision_cmd.go [hunk 7]
```diff
@@ -79,6 +84,11 @@ func NewProvisionCmd(cxt *command.Context) *cobra.Command {
 		"Additional parameter, whose value is stored in a secret, to use when provisioning the service, format: SECRET[KEY]")
 	cmd.Flags().StringVar(&provisionCmd.jsonParams, "params-json", "",
 		"Additional parameters to use when provisioning the service, provided as a JSON object. Cannot be combined with --param")
+	cmd.Flags().BoolVar(&provisionCmd.wait, "wait", false,
```
Comment (n3wscott) [inline cmd/svcat/instance/provision_cmd.go]: Why not bump these out to the top level flags? are there commands where this does not apply? Or a common function to call to add to cmd.Flags() to mix-in the wait flags.

Comment (carolynvs) [reply cmd/svcat/instance/provision_cmd.go]: I don't think it should be a top level flag because --wait for `svcat get plans` for example won't be implemented.

I will do the same thing I did for namespace's helper functions, so that there is a common interface for commands that support waiting.

Comment (n3wscott) [reply cmd/svcat/instance/provision_cmd.go]: if the broker is just a proxy, even get plans might take awhile... just saying OSB is flexible.

Comment (carolynvs) [reply cmd/svcat/instance/provision_cmd.go]: When I run `svcat get plans`, I am hitting the service catalog api, not the broker, so if the broker isn't finished listing (because it's slow or something), then I get nothing back. If I wanted to wait on something I would use `svcat get broker slowassbroker --wait` until it flips from an empty status/errfetchingcatalog to ready.

The idea behind wait is to wait for a mutating operation on a resource to complete (successfully or otherwise). I'm not sure that it applies to idempotent list operations. If we wanted something like that, I'd want to hash out what the UX would be and probably use a differently named flag for that. You know?

**Comment by jberkhahn** [comment]
This looks neat. You should add a test in each of the files in `pkg/svcat/service-catalog` for the new methods you added. Hopefully that isn't too hard to figure out, if not I could pair with you and write some ginkgo tests.

**Comment by n3wscott** [comment]
`--interval` looks nice, thanks for adding that! LGTM

**Comment by n3wscott** [comment]
fix the test before merging if that error is real...
